### PR TITLE
connection add CLI - validate the input params

### LINKED ISSUE #22392: Unknown connection types fail in cryptic ways
### Apache Airflow version

2.2.4 (latest released)

### What happened

I created a connection like:
```
airflow connections add fsconn --conn-host /tmp --conn-type File
```

When I really should have created it like:
```
airflow connections add fsconn --conn-host /tmp --conn-type fs
```

While using this connection, I found that FileSensor would only work if I provided absolute paths.  Relative paths would cause the sensor to timeout because it couldn't find the file.  Using `fs` instead of `File` caused the FileSensor to start working like I expected.

### What you think should happen instead

Ideally I'd have gotten an error when I tried to create the connection with an invalid type.

Or if that's not practical, then I should have gotten an error in the task logs when I tried to use the FileSensor with a connection of the wrong type.

### How to reproduce

_No response_

### Operating System

debian (in docker)

### Versions of Apache Airflow Providers

n/a

### Deployment

Astronomer

### Deployment details

`astro dev start`

### Anything else

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by Bowrna**
closes: #22392
relates: #22392 

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

**Review by potiuk — COMMENTED** [review]
We will need two tests:

* (parameterized) test covering valid and invalid connection types for _valid_conn_type
* CLI test checking the warning messages (you will find some CLI tests in the code - we have quite good coverage

**Comment by Bowrna** [comment]
> We will need two tests:
> 
> * (parameterized) test covering valid and invalid connection types for _valid_conn_type
> * CLI test checking the warning messages (you will find some CLI tests in the code - we have quite good coverage

@potiuk i will update this PR with tests

File: airflow/cli/commands/connection_command.py [hunk 1]
```diff
@@ -133,6 +134,24 @@ def _valid_uri(uri: str) -> bool:
     return uri_parts.scheme != '' and uri_parts.netloc != ''
 
 
+def _valid_conn_type(conn_type: str) -> Tuple[bool, List[str]]:
```
Comment (uranusjr) [inline airflow/cli/commands/connection_command.py]: Eagerly accessing valid conn types looks wrong to me (it’s slow). The `_get_connection_types()` function should live outside of this, and only called when `conn_type` is not valid.

This function should simply return a bool to indicate whether the conn type is valid.

**Review by uranusjr — DISMISSED** [review]

**Comment by Bowrna** [comment]
@uranusjr I have made the changes requested. Even if i move the `_get_connection_types()` function outside, i have to invoke that function to get the list of valid connection type. I could determine if the `conn-type` is valid or not after invoking  `_get_connection_types()` function. Could you explain to me if I am missing to understand any point you mentioned?

**Comment by uranusjr** [comment]
Add a `@cache` decorator on the function

File: airflow/cli/commands/connection_command.py [hunk 2]
```diff
@@ -195,6 +212,16 @@ def connections_add(args):
...
         raise SystemExit('Cannot supply both conn-uri and conn-json')
 
+    if has_type:
+        is_valid_conn_type = _valid_conn_type(args.conn_type)
+        if not is_valid_conn_type:
+            warnings.warn(f'The type provided to --conn-type is invalid: {args.conn_type}')
+            warnings.warn(
+                f'Supported --conn-types are:{_get_connection_types()}.'
+                'Hence overriding the conn-type with generic'
+            )
```
Comment (uranusjr) [inline airflow/cli/commands/connection_command.py]: Not sure `warnings.warn` is the right thing to do (not faimilar with how commands log things). Maybe `log.warning` is better?

**Comment by Bowrna** [comment]
> Add a `@cache` decorator on the function

Yes that makes sense. I will add it @uranusjr

**Comment by Bowrna** [comment]
@uranusjr Regarding warnings.warn I got the warning message when i tried to execute the command with invalid `conn-type`. Do I still have to use `log.warning`?

**Comment by uranusjr** [comment]
I’m not sure, to be honest, maybe another maintainer can offer better advice.

File: airflow/cli/commands/connection_command.py [hunk 3]
```diff
@@ -133,6 +135,25 @@ def _valid_uri(uri: str) -> bool:
     return uri_parts.scheme != '' and uri_parts.netloc != ''
 
 
+@lru_cache(maxsize=None)
```
Comment (uranusjr) [inline airflow/cli/commands/connection_command.py]: Use `@cache` from `airflow.compat` instead

File: airflow/cli/commands/connection_command.py [hunk 4]
```diff
@@ -133,6 +135,25 @@ def _valid_uri(uri: str) -> bool:
...
+            _connection_types.append(connection_type)
+    return _connection_types
+
+
+def _valid_conn_type(conn_type: str) -> bool:
+    is_valid = False
+    conn_types = _get_connection_types()
+    if conn_type in conn_types:
+        is_valid = True
+    return is_valid
```
Comment (uranusjr) [inline airflow/cli/commands/connection_command.py]:
```suggestion
def _valid_conn_type(conn_type: str) -> bool:
    return conn_type in _get_connection_types()
```

(Do we still need this function now it’s simplied to just one line?)

**Comment by Bowrna** [comment]
Do i have to make any other improvements in this code to get it merged? :)

**Review by uranusjr — APPROVED** [review]

**Review by potiuk — APPROVED** [review]

**Pull request merged**